src/pandas_openscm/index_manipulation.py

```diff
@@ -714,3 +717,188 @@ def update_levels_from_other(
     res = pd.MultiIndex(levels=levels, codes=codes, names=names)
 
     return res
+
+
+def create_level_from_collection(
+    level: str, value: Collection[Any]
+) -> tuple[pandas.Index[Any], list[int]]:
+    """
+    Create new level and corresponding codes.
+
+    Parameters
+    ----------
+    level
+        Name of the level to create
+
+    value
+        Values to use to create the level
+
+    Returns
+    -------
+    :
+        New level and corresponding codes
+    """
+    new_level: pandas.Index[Any] = pd.Index(value, name=level)
+    if not new_level.has_duplicates:
+        # Fast route, can just return new level and codes from level we mapped from
+        return value, list(range(len(value)))
+    # Slow route, have to update the codes
+    new_level = new_level.unique()
+    new_codes = new_level.get_indexer(value)  # type: ignore
+
+    return new_level, new_codes
+
+
+def set_levels(
+    ini: pd.MultiIndex, levels_to_set: dict[str, Any | Collection[Any]]
+) -> pd.MultiIndex:
+    """
+    Set the levels of a MultiIndex to the provided values
+
+    Parameters
+    ----------
+    ini
+        Input MultiIndex
+
+    levels_to_set
+        Mapping of level names to values to set. If values is of type `Collection`,
+        it must be of the same length as the MultiIndex. If it is not a `Collection`,
+        it will be set to the same value for all levels.
+
+    Returns
+    -------
+    :
+        New MultiIndex with the levels set to the provided values
+
+    Raises
+    ------
+        TypeError
+            If `ini` is not a MultiIndex
+        ValueError
+            If the length of the values is not equal to the length of the index
```

Comment:
```suggestion
    TypeError
        If `ini` is not a MultiIndex
    ValueError
        If the length of the values is a collection that is not equal to the length of the index
```

No idea why, but these bits aren't indented